src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Web/ProjectTypeGuidsDataSource.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Microsoft.VisualStudio.Text;
+
+using GuidCollectionProjectValue = Microsoft.VisualStudio.ProjectSystem.IProjectVersionedValue<System.Collections.Immutable.IImmutableList<System.Guid>>;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Web
+{
+    /// <summary>
+    ///     Provides project type <see cref="Guid"/> instances from the ProjectTypeGuids property.
+    /// </summary>
+    [Export]
+    internal class ProjectTypeGuidsDataSource : ChainedProjectValueDataSourceBase<IImmutableList<Guid>>
+    {
+        private readonly IProjectSnapshotService _snapshotService;
+        private string? _lastValue;
+
+        [ImportingConstructor]
+        public ProjectTypeGuidsDataSource(UnconfiguredProject project, IProjectSnapshotService snapshotService)
+            : base(project, synchronousDisposal: true)
+        {
+            ItemTypeGuidProviders = new OrderPrecedenceImportCollection<IItemTypeGuidProvider>(projectCapabilityCheckProvider: project);
+
+            _snapshotService = snapshotService;
+        }
+
+        [ImportMany]
+        private OrderPrecedenceImportCollection<IItemTypeGuidProvider> ItemTypeGuidProviders { get; }
+
+        public async Task<IImmutableList<Guid>> GetProjectTypeGuids()
+        {
+            using (JoinableCollection.Join())
+            {
+                GuidCollectionProjectValue snapshot = await SourceBlock.ReceiveAsync();
+
+                return snapshot.Value;
+            }
+        }
+
+        protected override IDisposable? LinkExternalInput(ITargetBlock<GuidCollectionProjectValue> targetBlock)
+        {
+            JoinUpstreamDataSources(_snapshotService);
+
+            DisposableValue<ISourceBlock<GuidCollectionProjectValue>> block = 
+                _snapshotService.SourceBlock.TransformManyWithNoDelta(CreateSnapshot);
+
+            block.Value.LinkTo(targetBlock, DataflowOption.PropagateCompletion);
+
+            return block;
+        }
+
+        private IEnumerable<GuidCollectionProjectValue> CreateSnapshot(IProjectVersionedValue<IProjectSnapshot> snapshot)
+        {
+            string value = snapshot.Value.ProjectInstance.GetPropertyValue(ConfigurationGeneral.ProjectTypeGuidsProperty);
+
+            if (StringComparers.PropertyValues.Equals(value, _lastValue))
```

Comment:
I guess this check is here, because this function creates a difference between the current and the previous snapshots?

Reply:
This was to avoid unnecessarily posting an update on no changes, but this ended up causing the hang because downstream a component was sync-linking the result of ProjectTypeCapabilitiesProvider and wasn't seeing version-only changes.